zio-http/src/main/scala/zhttp/service/Handler.scala

```diff
@@ -106,87 +107,84 @@ private[zhttp] final case class Handler[R](
               case Some(cause) =>
                 UIO {
-                  unsafeWriteAndFlushErrorResponse(cause)
-                  releaseRequest(jReq)
+                  processErrorResponse(jReq, Some(cause))
                 }
               case None        =>
                 UIO {
-                  unsafeWriteAndFlushEmptyResponse()
-                  releaseRequest(jReq)
+                  processErrorResponse(jReq, None)
                 }
             },
             res =>
-              if (self.isWebSocket(res)) UIO(self.upgradeToWebSocket(ctx, jReq, res))
-              else {
-                for {
-                  _ <- UIO { unsafeWriteAnyResponse(res) }
-                  _ <- res.data match {
-                    case HttpData.Empty =>
-                      UIO {
-                        unsafeWriteAndFlushLastEmptyContent()
-                      }
-
-                    case data @ HttpData.Text(_, _) =>
-                      UIO {
-                        unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
-                      }
-
-                    case HttpData.BinaryByteBuf(data) =>
-                      UIO {
-                        unsafeWriteLastContent(data)
-                      }
-
-                    case data @ HttpData.BinaryChunk(_) =>
-                      UIO {
-                        unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
-                      }
-
-                    case HttpData.BinaryStream(stream) =>
-                      writeStreamContent(stream)
-                  }
-                  _ <- Task(releaseRequest(jReq))
-                } yield ()
+              UIO {
+                processResponse(jReq, res)
               },
           )
         }
 
       case HExit.Success(res) =>
-        if (self.isWebSocket(res)) {
-          self.upgradeToWebSocket(ctx, jReq, res)
-        } else {
-          unsafeWriteAnyResponse(res)
-
-          res.data match {
-            case HttpData.Empty =>
-              unsafeWriteAndFlushLastEmptyContent()
-              releaseRequest(jReq)
-
-            case data @ HttpData.Text(_, _) =>
-              unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
-              releaseRequest(jReq)
-
-            case HttpData.BinaryByteBuf(data) =>
-              unsafeWriteLastContent(data)
-              releaseRequest(jReq)
-
-            case data @ HttpData.BinaryChunk(_) =>
-              unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
-              releaseRequest(jReq)
-
-            case HttpData.BinaryStream(stream) =>
-              unsafeRunZIO(
-                writeStreamContent(stream) *> Task(releaseRequest(jReq)),
-              )
-          }
-        }
+        processResponse(jReq, res)
+      case HExit.Failure(e)   =>
+        processErrorResponse(jReq, Some(e))
+      case HExit.Empty        =>
+        processErrorResponse(jReq, None)
+    }
+
```

Comment:
Had to refactor as there was the same (success / error) response  flow with just one difference of effect and non effect computation.

Reply:
Don't allocate additional `Some` along the way. Slowly such allocations start affecting performance quite badly.